apteryx.c

```diff
@@ -201,6 +202,72 @@ handle_index (rpc_message msg)
     return true;
 }
 
+struct delayed_watch {
+    gpointer handle;
+    uint64_t ref;
+    void *fn;
+    GNode *root;
+    void *data;
+    guint timeout_ms;
+};
+static GList *dw_list = NULL;
+
+static gboolean
+dw_process (gpointer arg1)
+{
+    struct delayed_watch *dw = (struct delayed_watch *) arg1;
+
+    dw_list = g_list_remove (dw_list, dw);
+
+    DEBUG ("WATCH-TREE(delayed) CB (0x%"PRIx64")\n", dw->ref);
+    if (apteryx_debug) {
+        apteryx_print_tree (dw->root, stdout);
+    }
+    if (dw->data)
+        ((void*(*)(const GNode*, void*)) dw->fn) (dw->root, dw->data);
+    else
+        ((void*(*)(const GNode*)) dw->fn) (dw->root);
+    g_free (dw);
+    return G_SOURCE_REMOVE;
+}
+
+static void
+dw_add (uint64_t ref, void *fn, GNode *root, void *data, guint timeout_ms)
+{
+    struct delayed_watch *dw = NULL;
+
+    /* Find existing delayed watch */
+    for (GList *iter = dw_list; iter; iter = g_list_next (iter))
+    {
+        dw = (struct delayed_watch *) iter->data;
+        if (dw->ref == ref && dw->fn == fn && dw->data == data && dw->timeout_ms == timeout_ms)
+            break;
+        dw = NULL;
+    }
+
+    /* If we found a match lets merge in the new data */
+    if (dw)
+    {
+        dw_list = g_list_remove (dw_list, dw);
+        rpc_del_callback (rpc, dw->handle);
+        root = apteryx_merge_tree (dw->root, root);
+        g_free (dw);
+        dw = NULL;
+    }
+
+    /* Create a new timer for the requested timeout */
+    if (!dw)
```

Comment:
Just a plain "else" ? I think if we merged the match in above we will then create another new one here...